include/pybind11/cast.h

```diff
@@ -960,9 +960,15 @@ template <typename type> class type_caster<std::reference_wrapper<type>> {
 private:
     using caster_t = make_caster<type>;
     caster_t subcaster;
-    using subcaster_cast_op_type = typename caster_t::template cast_op_type<type>;
-    static_assert(std::is_same<typename std::remove_const<type>::type &, subcaster_cast_op_type>::value,
-            "std::reference_wrapper<T> caster requires T to have a caster with an `T &` operator");
```

Comment:
> > > (except for the newly added `test_freezable_type_caster.cpp` that I haven't yet looked at)
> > 
> > 
> > Additionally, the only thing stopping `test_freezable_type_caster` from compiling/passing is the `static_assert`, which seems to indicate to me that something's the matter with the `type_caster`?
> 
> Which `static_assert`, precisely? The one in the `type_caster<std::reference<T>>` specialization? That's the specific change that I made without which my test fails to build.

These 2 lines. If I'm not dreaming, all tests pass when I comment these out (in the original code, without your changes). At least on my Linux, Python 3.8, but this part of the casters should be independent from Python versions, I believe.

Moreover, all tests (except for `test_freezable_type_caster`) pass with the original code.

Reply:
Exactly the point of the change.


1/ For std::reference<T>, the original implementation determines operator to call via the equivalent of `cast_op_type = make_type_caster<T>::cast_op<std::decay<T>>` and then invokes `operator cast_op_type&()`. Basically removing the reference and the const before getting the cast op, then adding the reference back in.

The original static_assert ensures that there is such an operator as `cast_op_type&` on `make_type_caster<T>`.


2/ The updated implementation determines the operator to call via the equivalent of `cast_op_type = make_type_caster<T>::cast_op<std::add_reference<T>>` and then invokes `operator cast_op_type()`. 
It adds the reference to the desired type before invoking cast_op, and then uses that operator without rewriting.

The static assert is the equivalent, requiring that the cast operator is one of either `T&` or `const T&`. The new behavior allows cast_op<> to select which operator overload to call instead of manipulating it afterwards, and the static_assert requires that the behavior is either the same as before (T&), or allows it to be const (const T&).

It's good that all the original tests pass with the original code; the fact that the added test fails also indicates that the code works as desired.

Reply:
> It's good that all the original tests pass with the original code; the fact that the added test fails also indicates that the code works as desired.

Nono, what I'm saying is that the new tests you added (except for `test_freezable_type_caster`) pass with the old code.
And `test_freezable_type_caster` also passes with the old code (ignoring the static_assert).

So either your tests are not demonstrating things well, or cannot see a good reason for your changes to `cast.h`. According to this experiment, this only thing that needs to be changed is a more relaxed `static_assert` check?

Apart from that, I need another fresh look tomorrow. I can see your 1/ and 2/, what's happening, but the way other casters are written, I'm far from convinced that that is the right way to go. `const` is just something that's currently not supported by pybind11 and would probably take a lot more work to support.

Meanwhile, do you have a single, simple example, that show that `std::reference_wrapper<T>` currently (on `master`) behaves differently from `T&`? (as stated in your title, "Allow type_caster of std::reference_wrapper<T> to behave like a native reference.")

Reply:
No, you're right. The `static_assert` is the most significant change. The `cast_op<add_reference<T>>` part of the change only matters when you do different things with cast_op<T>& and cast_op<T&>. And making a reasonable test that does that is challenging.

As for my other claim regarding const, it appears that on re-reading the it was not completely correct. I was originally using `cast_op_type`, which is where `intrinsic_t<T>`, which removes const, was applied, and the const-removal was not native to reference_wrapper.

The test now demonstrates the failure case better because the before behavior invokes the cast_op specialization in a way that discards const when not passed as a reference or a pointer, and the roundtrip tests now fail with the older variant beyond just the static_assert failure.